src/AggregateFunctions/AggregateFunctionDistinctDynamicTypes.cpp

```diff
@@ -0,0 +1,161 @@
+#include <unordered_set>
+#include <IO/WriteHelpers.h>
+#include <IO/ReadHelpers.h>
+
+#include <DataTypes/DataTypeString.h>
+#include <DataTypes/DataTypeArray.h>
+#include <DataTypes/DataTypesBinaryEncoding.h>
+#include <Columns/ColumnDynamic.h>
+
+#include <AggregateFunctions/IAggregateFunction.h>
+#include <AggregateFunctions/AggregateFunctionFactory.h>
+#include <AggregateFunctions/FactoryHelpers.h>
+
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
+    extern const int ILLEGAL_TYPE_OF_ARGUMENT;
+    extern const int TOO_LARGE_ARRAY_SIZE;
+}
+
+struct AggregateFunctionDistinctDynamicTypesData
+{
+    constexpr static size_t MAX_ARRAY_SIZE = 0xFFFFFF;
+
+    std::unordered_set<String> data;
+
+    void add(const String & type)
+    {
+        data.insert(type);
+    }
+
+    void merge(const AggregateFunctionDistinctDynamicTypesData & other)
+    {
+        data.insert(other.data.begin(), other.data.end());
+    }
+
+    void serialize(WriteBuffer & buf) const
+    {
+        writeVarUInt(data.size(), buf);
+        for (const auto & type : data)
+            writeStringBinary(type, buf);
+    }
+
+    void deserialize(ReadBuffer & buf)
+    {
+        size_t size;
+        readVarUInt(size, buf);
+        if (size > MAX_ARRAY_SIZE)
```

Comment:
This is a generic question. Shouldn't `size > MAX_ARRAY_SIZE` check be done inside readVarUInt? This is because this setting allows reading an array bigger than `MAX_ARRAY_SIZE`, and then testing the length. Could the process get OOM-killed in the worst case?

Reply:
`readVarUInt` reads only `size_t` value, it doesn't read any array values. We check for too big array size before reading array elements, there is no way to get OOM.